src/main/java/org/opensearch/knn/index/mapper/LuceneFieldMapper.java

```diff
@@ -63,9 +49,10 @@ public class LuceneFieldMapper extends KNNVectorFieldMapper {
 
         this.knnMethod = input.getKnnMethodContext();
         final SpaceType spaceType = this.knnMethod.getSpaceType();
-        final VectorSimilarityFunction vectorSimilarityFunction = Optional.ofNullable(
-            SPACE_TYPE_TO_VECTOR_SIMILARITY_FUNCTION.get(spaceType)
-        ).orElseThrow(() -> new IllegalArgumentException(String.format("Space type [%s] is not supported for Lucene engine", spaceType)));
+        final VectorSimilarityFunction vectorSimilarityFunction = Optional.ofNullable(spaceType.getVectorSimilarityFunction())
+            .orElseThrow(
+                () -> new IllegalArgumentException(String.format("Space type [%s] is not supported for Lucene engine", spaceType))
+            );
```

Comment:
spaceType.getVectorSimilarityFunction() will never return the null. If the vectorSimilarityFunction is not there then it is throwing the exception. So Optional.ofNullable is not doing anything here.

But you should definitely check if spaceType is null or not. That is a valid check to make.

Reply:
> spaceType.getVectorSimilarityFunction() will never return the null. If the vectorSimilarityFunction is not there then it is throwing the exception. So Optional.ofNullable is not doing anything here.

Makes sense. Will remove.

> But you should definitely check if spaceType is null or not. That is a valid check to make.

Good call out. I was thinking about this. In KNNMethodContext, these [3 variables](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L60-L62) are implicitly non-null. They are implicitly non-null because [reading from the stream](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L70) uses readString instead of readOptionalString and, in the [parse function](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L110) which uses the all args constructor, they are all initialized. That being said, Im wondering if it makes more sense to explicitly make them non-null here and add proper null checks, instead of checking in the mapper. What do you think?

Reply:
Here comes the catch with the AllArgsConstructor, that constructor is not guarantee that it take only NonNulls, it can take null if you see it separately.
But when you combine it with the parseFunction it make sure that [SpaceType value is not null](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L119), by setting a default value.

This is creating a dependency between one another. So, here are your options:
1. Make the field spaceType as NonNull always by using @NonNull annotation [here](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L61) and avoid explicit null checks.
2. You can have an explicit null check here in your code.

I would prefer 2 over 1, just because of flexibility, but if we feel that SpaceType can never be null in any case then go with 1.

Reply:
Right that makes sense. That being said, I prefer option 1. Given that the stream methods ([ctor](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L70) and [writeTo](https://github.com/opensearch-project/k-NN/blob/feature/lucene-knn/src/main/java/org/opensearch/knn/index/KNNMethodContext.java#L221)) would fail for null values, I think it makes sense to add the nonnull annotation. Additionally, I cant think of a case where it would be advantageous to allow these methods to be null.